inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/sidebar/LearningCurveChartPanel.java

```diff
@@ -18,164 +18,122 @@
 
-            // append data columns
-            dataColumns.append("', ");
-            dataColumns.append(data);
-            dataColumns.append("]");
-            dataColumns.append(",");
+        // iterate over recommenderScoreMap to create data
+        for (String recommenderName : recommenderScoreMap.keySet()) {
+            // extract the scores from the recommenderScoreMap. The format of data is a comma
+            // separated string of scores(each score is Double cast-able) to be. 
+            // Example 2.3, 4.5 ,6, 5, 3, 9,
+            String data = recommenderScoreMap.get(recommenderName).stream().map(Object::toString)
+                    .collect(Collectors.joining(", "));
+            
+            curveData.put(recommenderName,data);
+            
+            learningCurve.setCurveData(curveData);
+            
+            // the Curve is not allowed to have more points as compared to MAX_POINTS_TO_PLOT. This
+            // is how many scores we have retrieved from the database
+            learningCurve.setMaximumPointsToPlot(MAX_POINTS_TO_PLOT);
         }
- 
-        try {
-            String javascript = createJSScript(dataColumns.toString(), chartType.toString());
-            log.debug("Rendering Recommender Evaluation Chart: {}", javascript);
 
-            aEvent.getRequestHandler().prependJavaScript(javascript);
-        }
-        catch (IOException e) {
-            log.error("Unable to render chart", e);
-            error("Unable to render chart: " + e.getMessage());
-            aEvent.getRequestHandler().addChildren(getPage(), IFeedback.class);
-        }
-    }
-
-    /**
-     * Creates the JS script to render graph with the help of given data points. Also creates an
-     * x-axis of a sequence from 0 to maximumNumberOfPoints (50). Example value of
-     * aDataColumns: 
-     * <pre>
-     * ['recommender1', 1.0, 2.0, 3.0 ], ['recommender2', 2.0, 3.0, 4.0]
-     * </pre>
-     * 
-     * Example value of aChartType
-     * <pre>
-     * recommender1: 'step', recommender2 : 'step'
-     * </pre>
-     */
-    private String createJSScript(String aDataColumns, String aChartType) throws IOException
-    {
-        int[] intArray = IntStream.range(0, maxPointsToPlot).map(i -> i).toArray();
-        String xaxisValues = "[ 'x' ," + substring(Arrays.toString(intArray), 1, -1) + "]";
-        String data = toJsonString(aDataColumns).substring(1, aDataColumns.toString().length());
-
-        // bind data to chart container
-        String javascript = "var chart=c3.generate({bindto:'#" + chartContainer.getMarkupId()
-                + "',data:{ x:'x', columns:[" + xaxisValues + " ," + data + "],types:{" + aChartType
-                + "}},axis: { y : { tick : { format: function(d){return Math.round(d * 10000) / 10000}}}}});;";
-        return javascript;
+        //provide the chart above calculated data to plot the learning curve
+        chartPanel.setDefaultModel(Model.of(learningCurve));
```

Comment:
Instead of triggering the re-rendering in `onModelChanged()`, you can add the chartPanel to the request handler here, no?

Reply:
As in, move the following code here from the constructor?

```
ChartPanel chartPanel = new ChartPanel(MID_CHART_CONTAINER, Model.of());
        chartPanel.setOutputMarkupId(true);
        add(chartPanel);    
```

If I do not add the chart panel in the constructor then there will be an error because the html is looking for a `chart-container` component to be added to the panel.

Reply:
No. As in:

```
chartPanel.setDefaultModel(Model.of(learningCurve));
aRequestTarget.add(chartPanel);
```

Reply:
Then I must render the chart in `onRender()` of the `ChartPanel`?

The following returns `null` in the onRender() of `ChartPanel` and I need a request target object to render chart.
``` RequestCycle.get().find(AjaxRequestTarget.class)```

Reply:
I believe that *if* `onRender` is called as part of an AJAX request, then `RequestCycle.get().find(AjaxRequestTarget.class)` should not return null. But it is of course null if the rendering happens outside an AJAX request (e.g. in a full page reload).

Reply:
Isn't `renderHead()` only called once when the component is added?  I believe `renderHead()` is not called every time the component is added to an AJAX target. 

To be honest, I am having the similar problems in other places where JavaScript is used (e.g. the annotation editor plugins). Optimally, the JS needs to be executed in two situations: 1) on a page reload, 2) when the component is added to an AJAX target. But I haven't (yet) figured out the best way to do either. If for 1) the JS is added to `renderHead()`, then it ends up in the page source code - that seems be a bit annoying when sending (large) JSON data. For 2), a good place to attach the JS to the AJAX target needs to be found (maybe `render()` or probable better `onRender()` or `onAfterRender()`).

Because I do atm not have good solutions for these problems, the `BratAnnotationEditor` does things a bit differently. The first case (page reload) is solved by adding a short JS to the header which triggers an AJAX request that loads the actual data. This way, I can avoid the whole data ending up in the page source code. For the second case, instead of adding the editor to the AJAX target (`target.add(editor)`), I am calling the method `editor.requestRender(target)` which schedules a `RenderListener` to be executed at the very end of the AJAX request - this listener then generates the JS and attaches it to the AJAX request. 

If we find a good solution that allows using `target.add(componentUsingJs)`, that would be great!

Reply:
> Isn't renderHead() only called once when the component is added? I believe renderHead() is not called every time the component is added to an AJAX target.

`renderHead()` is called every time the component is added to the AJAX target which is how the chart updates on every annotation event.

> To be honest, I am having the similar problems in other places where JavaScript is used (e.g. the annotation editor plugins). Optimally, the JS needs to be executed in two situations: 1) on a page reload, 2) when the component is added to an AJAX target. But I haven't (yet) figured out the best way to do either. If for 1) the JS is added to renderHead(), then it ends up in the page source code - that seems be a bit annoying when sending (large) JSON data. For 2), a good place to attach the JS to the AJAX target needs to be found (maybe render() or probable better onRender() or onAfterRender()).

If I understood it correct, I must look for a way to move the code from `renderHead` of the `ChartPanel` to onRender() or something else to avoid making the javascript a part of the page source code irrespective of whether it is functioning correctly?

Reply:
I am not aware of a way of adding JS in `onRender()` unless you would be in an AJAX request. I think that the Chart needs to be initialized in `renderHead()` but I would tend to avoid including the data already in the initialization. If possible, I would suggest expose the data through an `AbstractAjaxBehavior` (cf. e.g. `PdfAnnoPanel pdftxtProvider`) and to send a JS snippet in `renderHead()` which triggers the chart to load the data.

No regarding rendering as part of an AJAX request via `render()` or `onRender()`. If you add a component to an AJAX request, that normally causes the HTML of the component to be removed and to be re-added. This destroys the chart which was rendered using JS. So either the chart needs to be reinitialized (i.e. using the same JS code used in `renderHead()`) or we would have to try to intercept the replacement somehow and instead of removing/re-adding the HTML, we would just send a JS to tell the chart to update itself. However, I have no idea if that is possible. The alternative would be not to add the chart component to the AJAX target (thus avoiding the remove/re-add) and instead introduce a new method like `render(target)` which sends a JS to the chart reload its data - and then we'd call that. Makes more sense now?

Reply:
If you have any other ideas, let me know.

Reply:
I got your points. I am looking into the options. Just to be clear about the problem, the goal is to not have potentially large javascript in the source code of the page. Right?

Reply:
>  I would suggest expose the data through an AbstractAjaxBehavior and to send a JS snippet in renderHead() which triggers the chart to load the data.

I tries this. Couldn't find a better way.